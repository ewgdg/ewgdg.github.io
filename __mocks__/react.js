const React = jest.requireActual("react")
module.exports = {
  ...React,
  useContext: jest.fn().mockImplementation(React.useContext),
}
