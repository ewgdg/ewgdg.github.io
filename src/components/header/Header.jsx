/* eslint-disable react/destructuring-assignment */

import PropTypes from "prop-types"
import React, { useContext, useEffect, useState } from "react"
import AppBar from "@material-ui/core/AppBar"
import Button from "@material-ui/core/Button"
import Toolbar from "@material-ui/core/Toolbar"
import IconButton from "@material-ui/core/IconButton"

import Typography from "@material-ui/core/Typography"
import Box from "@material-ui/core/Box"

import Fab from "@material-ui/core/Fab"
import KeyboardArrowUpIcon from "@material-ui/icons/KeyboardArrowUp"
import Zoom from "@material-ui/core/Zoom"

import { navigate } from "gatsby"

import LayoutContext from "../../contexts/LayoutContext"
import { scrollIntoView, ScrollDetector } from "../../utilities/scroll"
import { clearHistoryState } from "../../contexts/useRestoreComponentState"
import useLayoutContext from "../../contexts/useLayoutContext"

function useScrollTrigger({ threshold, target }) {
  const [trigger, setTrigger] = useState(false)

  useEffect(() => {
    const scene = new ScrollDetector({
      scrollLayer: target,
      triggerHook: 0,
      offset: threshold,
    })

    scene.setEventListener(progress => {
      if (progress > 0) {
        setTrigger(true)
      } else {
        setTrigger(false)
      }
    })

    return () => {
      scene.destroy()
      setTrigger(false)
    }
  }, [threshold, target])
  return trigger
}
function BackToTop(props) {
  const { children, anchorId } = props
  // const classes = useStyles();
  const context = useContext(LayoutContext)
  const trigger = useScrollTrigger({
    threshold: 100,
    target: context.scrollLayer,
  })

  const handleClick = event => {
    const anchor = (event.target.ownerDocument || document).querySelector(
      `#${anchorId}`
    )
    if (anchor) {
      scrollIntoView(anchor, context.scrollLayer, 888)
      // anchor.scrollIntoView({ behavior: "smooth", block: "start" })
    }
  }

  return (
    <Zoom in={trigger}>
      <div onClick={handleClick} role="presentation" className="scroll-button">
        {children}
        <style jsx>
          {`
            .scroll-button {
              position: fixed;
              bottom: 2rem;
              right: 2rem;
            }
          `}
        </style>
      </div>
    </Zoom>
  )
}
BackToTop.propTypes = {
  children: PropTypes.element.isRequired,
  anchorId: PropTypes.string.isRequired,
}

const Header = ({ position, color, style }) => {
  const context = useLayoutContext()
  return (
    <div>
      <AppBar
        position={position}
        id="navbar-top"
        className="navbar"
        style={style}
      >
        <Toolbar style={{ justifyContent: "space-around" }} variant="dense">
          <IconButton
            edge="start"
            color="inherit"
            aria-label="menu"
            onClick={() => {
              clearHistoryState(["/"], context)
              navigate("/")
            }}
          >
            {/* <MenuIcon /> */}
            <Typography variant="h6" className="title">
              Xian
            </Typography>
          </IconButton>

          <Button
            onClick={() => {
              navigate("/about")
            }}
            color="inherit"
          >
            <Box display="flex" flexDirection="row">
              <Typography variant="h6" className="title">
                About
              </Typography>
            </Box>
          </Button>

          <Button
            onClick={() => {
              clearHistoryState(["/blog"], context)
              navigate("/blog")
            }}
            color="inherit"
          >
            Blog
          </Button>
        </Toolbar>

        <BackToTop anchorId="navbar-top">
          <Fab color="secondary" size="small" aria-label="scroll back to top">
            <KeyboardArrowUpIcon />
          </Fab>
        </BackToTop>
      </AppBar>
      <style jsx global>
        {`
          .navbar {
            background-color: transparent;
            color: ${color};
            user-select: none;
          }
        `}
      </style>
    </div>
  )
}

Header.propTypes = {
  position: PropTypes.string,
  color: PropTypes.string,
}
Header.defaultProps = {
  position: "static",
  color: "black",
}
export default Header
