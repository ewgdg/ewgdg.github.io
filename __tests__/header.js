/* eslint-disable react/jsx-filename-extension */
import React from "react"
import renderer from "react-test-renderer"
import { render, fireEvent } from "@testing-library/react"
import Header from "../src/components/header/Header"

describe("Header", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<Header position="absolute" />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("renders correctly with testing library", () => {
    const { getByText } = render(<Header />)
    fireEvent.click(getByText(/about/i))
    expect(window.location.href).toBe("/about")
  })
})
