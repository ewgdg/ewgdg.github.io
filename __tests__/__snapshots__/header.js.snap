// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
<div>
  <header
    className="MuiPaper-root MuiPaper-elevation4 MuiAppBar-root MuiAppBar-positionAbsolute navbar MuiAppBar-colorPrimary"
    id="navbar-top"
    style={Object {}}
  >
    <div
      className="MuiToolbar-root MuiToolbar-dense MuiToolbar-gutters"
      style={
        Object {
          "justifyContent": "space-around",
        }
      }
    >
      <button
        aria-label="menu"
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <h6
            className="MuiTypography-root title MuiTypography-h6"
          >
            Xian
          </h6>
        </span>
      </button>
      <button
        aria-label="menu"
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <div
            className="MuiBox-root MuiBox-root-84"
          >
            <h6
              className="MuiTypography-root title MuiTypography-h6"
            >
              About
            </h6>
          </div>
        </span>
      </button>
      <button
        aria-label="menu"
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <h6
            className="MuiTypography-root title MuiTypography-h6"
          >
            Blog
          </h6>
        </span>
      </button>
    </div>
    <div
      className="scroll-button"
      onClick={[Function]}
      role="presentation"
      style={
        Object {
          "transform": "scale(0)",
          "visibility": "hidden",
        }
      }
    >
      <button
        aria-label="scroll back to top"
        className="MuiButtonBase-root MuiFab-root MuiFab-sizeSmall MuiFab-secondary"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiFab-label"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z"
            />
          </svg>
        </span>
      </button>
      <style>
        
            .scroll-button {
              position: fixed;
              bottom: 2rem;
              right: 2rem;
            }
          
      </style>
    </div>
  </header>
  <style>
    
          .navbar {
            background-color: transparent;
            color: black;
            user-select: none;
          }
        
  </style>
</div>
`;
